File: bonus_content/tik_tok_effect_for_children/tik_tok_effect_src/Emoji Puzzles 4 Kids/Graph/ScriptTemplate/Subtract/CGSubtract.js
/**
 * @file CGSubtract.js
 * @author liujiacheng
 * @date 2021/8/23
 * @brief CGSubtract.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const {BaseNode} = require('./BaseNode');
const Amaz = effect.Amaz;

class CGSubtract extends BaseNode {
  constructor() {
    super();
  }

  setNext(index, func) {
    this.nexts[index] = func;
  }

  setInput(index, func) {
    this.inputs[index] = func;
  }

  getOutput() {
    let v1 = this.inputs[0]();
    let v2 = this.inputs[1]();
    if (v1 !== undefined && v2 !== undefined) {
      if (this.valueType === 'Vector2f') {
        return new Amaz.Vector2f(v1.x - v2.x, v1.y - v2.y);
      } else if (this.valueType === 'Vector3f') {
        return new Amaz.Vector3f(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z);
      } else if (this.valueType === 'Vector4f') {
        return new Amaz.Vector4f(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z, v1.w - v2.w);
      } else {
        return v1 - v2;
      }
    }
    return undefined;
  }
}

exports.CGSubtract = CGSubtract;

File: bonus_content/tik_tok_effect_for_children/tik_tok_effect_src/Emoji Puzzles 4 Kids/Graph/ScriptTemplate/Subtract/CGSubtract.lua
local CGSubtract = CGSubtract or {}
CGSubtract.__index = CGSubtract

function CGSubtract.new()
    local self = setmetatable({}, CGSubtract)
    self.inputs = {}
    return self
end

function CGSubtract:setInput(index, func)
    self.inputs[index] = func
end

function CGSubtract:getOutput(index)
    local v1 = self.inputs[0]()
    local v2 = self.inputs[1]()
    if v1 == nil or v2 == nil then
        return nil
    end
    return self.inputs[0]() - self.inputs[1]()
end

return CGSubtract
